### PR TITLE
special-case ExternalStream in bytes starts-with

### PR DESCRIPTION
**Pull request by Dorumin**
# Description
`bytes starts-with` converts the input into a `Value` before running .starts_with to find if the binary matches. This has two side effects: it makes the code simpler, only dealing in whole values, and simplifying a lot of input pipeline handling and value transforming it would otherwise have to do. _Especially_ in the presence of a cell path to drill into. It also makes buffers the entire input into memory, which can take up a lot of memory when dealing with large files, especially if you only want to check the first few bytes (like for a magic number).

This PR adds a special branch on PipelineData::ExternalStream with a streaming version of starts_with.

# User-Facing Changes
Opening large files and running bytes starts-with on them will not take a long time.

# Tests + Formatting

Don't forget to add tests that cover your changes.

Make sure you've run and fixed any issues with these commands:

- `cargo fmt --all -- --check` to check standard code formatting (`cargo fmt --all` applies these changes)
- `cargo clippy --workspace -- -D warnings -D clippy::unwrap_used -A clippy::needless_collect` to check that you're using the standard code style
- `cargo test --workspace` to check that all tests pass

# Drawbacks
Streaming checking is more complicated, and there may be bugs. I tested it with multiple chunks with string data and binary data and it seems to work alright up to 8k and over bytes, though.

The existing `operate` method still exists because the way it handles cell paths and values is complicated. This causes some "code duplication", or at least some intent duplication, between the value code and the streaming code. This might be worthwhile considering the performance gains (approaching infinity on larger inputs).

Another thing to consider is that my ExternalStream branch considers string data as valid input. The operate branch only parses Binary values, so it would fail. `open` is kind of unpredictable on whether it returns string data or binary data, even when passing `--raw`. I think this can be a problem but not really one I'm trying to tackle in this PR, so, it's worth considering.

### CODE REVIEW

**Comment by rgwood** [comment]
Thank you for the PR! There's a lot of work to do to improve streaming in Nushell and I really appreciate PRs that push this area forward.

I'll try to give this a close look today or tomorrow.

File: crates/nu-command/src/bytes/starts_with.rs [hunk 1]
```diff
@@ -60,13 +61,73 @@ impl Command for BytesStartsWith {
...
-            starts_with,
-            arg,
-            input,
-            call.head,
-            engine_state.ctrlc.clone(),
-        )
+
+        match input {
+            PipelineData::ExternalStream { stdout, span, .. } if stdout.is_some() => {
+                let stream = stdout.expect("stdout is some");
```
Comment (rgwood) [inline crates/nu-command/src/bytes/starts_with.rs:67]:
```suggestion
        match input {
            PipelineData::ExternalStream { stdout: Some(stream), span, .. } => {
```

**Comment by sholderbach** [comment]
To add tests of `PipelineData::ExternalStream` you can take a look at other tests that use `nu --testbin` to have a reproducible output from an external.

Stuff like that should be covered in full in my opinion.

**Comment by Dorumin** [comment]
@sholderbach Sounds reasonable. I was thinking of testing but there weren't any existing bytes starts-with tests, or lots of streaming tests that I could use as reference.

I've been looking at `--testbin` search results in the repository but I'm slightly more confused than before. How would I, for example, generate an `n` bytes output of a particular pattern as an ExternalStream? I was thinking of writing files to achieve that goal.

Some test cases I had in mind were:
- Files smaller than 8kb (the size that ExternalStream seems to chunk at for me), binary and string data. Easily reproducible, like ascending bytes from 0 to 255, and ascii text from 65 to 97.
- Files slightly larger than 8kb, so they're output in two or three chunks. These chunks may be mixed between binary and string data, I'm not sure how `open` maintains this, and if the same Value type is always output from a stream
- Simple binary tests that aren't streaming, although those are mostly handled by example tests

File: crates/nu-command/src/bytes/starts_with.rs [hunk 2]
```diff
@@ -60,13 +61,73 @@ impl Command for BytesStartsWith {
...
+
+                let mut i = 0;
+
+                for item in stream {
+                    match item {
+                        Ok(v) => match v {
+                            Value::String { val, span } => {
```
Comment (rgwood) [inline crates/nu-command/src/bytes/starts_with.rs:74]: I found this double-nested match a little hard to follow, and it would be nice to eliminate the duplicated code in the `Value::String` and `Value::Binary` branches. What about something like this instead?

```rust
let bytes: Vec<u8> = match item {
    Ok(Value::String { val, span }) => val.as_bytes().to_vec(),
    Ok(Value::Binary { val, span }) => val,
    Ok(Value::Error { error }) => todo!(), // early return
    Ok(other) => todo!(), // early return
    Err(err) => return Err(err), // early return
};
// check whether `bytes` matches` the remaining bytes in `pattern`
````

**Review by rgwood — COMMENTED** [review]
Thanks for the PR!

I think we'll need some tests to cover the new logic; it looks like it's currently not covered, but I'm not 100% confident in our code coverage tooling so just let me know if that's wrong.

File: crates/nu-command/src/bytes/starts_with.rs (hunk 1 above)
Comment (Dorumin) [reply crates/nu-command/src/bytes/starts_with.rs]: Excellent pattern matching, this makes me feel much better about things. Thanks.

File: crates/nu-command/src/bytes/starts_with.rs (hunk 2 above)
Comment (Dorumin) [reply crates/nu-command/src/bytes/starts_with.rs]: Implemented this suggestion and it does reduce code duplication in my branches. It's better, thanks!

I chose not to collect into a Vec<u8> in the string case, and instead matched on &item so the byte slice would live as long as `item`. This means that the cloning is required on the error case, though, which sounds like a reasonable tradeoff.

**Comment by Dorumin** [comment]
Added some tests, let's see where code coverage is at right now (and if they even pass in CI).

On that note, any easy way of checking code coverage locally on my branch changes without having to push first?

**Comment by Dorumin** [comment]
I'm not seeing any code coverage warnings so I guess that's a good thing, or maybe it broke

**Review by rgwood — APPROVED** [review]
This looks great; thanks for addressing the comments quickly, I'm really impressed with the tests.

**Comment by sholderbach** [comment]
Awesome thanks for the quick iteration! All relevant codepaths are covered so seems good to me as well!

**Comment by Dorumin** [comment]
I couldn't figure out how to get ExternalStream to return other kinds of data (it was a pain to even get it to spit Binary, even null bytes are valid Value::Strings, so I had to get explicitly invalid unicode sequences), so errors handling and the like is not well covered. Still, it type checks so it probably short circuits the errors just fine.

I might look into other commands and see if they buffer their input and could be converted to streaming later

**Comment by sholderbach** [comment]
> I couldn't figure out how to get ExternalStream to return other kinds of data (it was a pain to even get it to spit Binary, even null bytes are valid Value::Strings, so I had to get explicitly invalid unicode sequences), so errors handling and the like is not well covered. Still, it type checks so it probably short circuits the errors just fine.

Might possibly unreachable with what we have now. But having the defensive error handling in place should keep us safe in the future.

> I might look into other commands and see if they buffer their input and could be converted to streaming later

Awesome!

**Comment by rgwood** [comment]
@Dorumin If you're looking for (or creating) streaming issues we have a label: https://github.com/nushell/nushell/issues?q=is%3Aopen+is%3Aissue+label%3Astreaming
